README.md
# ahh - Andrew Huang Helps? v1.0.3

### Functions that I can easily reference, and maybe you too!

## HOW TO GET IT:
New-school method:
1. `pip install ahh`

2. Ensure your packages version (`pip list`) match with ones listed in requirements.txt

Old-school method:
1. Type `git clone https://github.com/ahuang11/ahh.git`

2. Go into ahh folder (where setup.py is)

3. Type `pip install -e .` (may need to be in bash first!)

4. In a Python script, type `from ahh import pre, era, vis, sci, ext`

## MOTIVATION TO GET STARTED:

#### Perhaps you've wrote these lines one too many times just to create a map that looks like this.
```python
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import cartopy.feature as cfeature

fig = plt.figure(figsize=(12, 8))
projection = ccrs.PlateCarree()
ax = plt.axes(projection=projection)
im = ax.contourf(da.lon, da.lat, da[0].values, transform=projection, cmap='RdBu_r')
_ = plt.colorbar(im, orientation='horizontal', shrink=0.5, pad=0.05)
_ = ax.gridlines(crs=ccrs.PlateCarree(), draw_labels=True)
_ = ax.coastlines()
plt.grid()
```
![Before Map](/readme_images/map_before.png)

#### With this package, it's simplified down to just this (and a prettier map)!
```python
from ahh import vis

ax = vis.plot_map(da[0], da.lat, da.lon)
```
![After Map](/readme_images/map_after.png)

#### Or perhaps you don't deal too much with maps, but you do a lot with timeseries plots.
```python
import matplotlib.pyplot as plt

plt.plot(x, y, label='Blue', linestyle='--')
plt.legend() # equivalent to below
```
![Before Map](/readme_images/timeseries_before.png)

#### Compare that with this package's version!
```python
from ahh import vis

vis.plot_line(x, y, label='Red', linestyle='--')
```
![After Map](/readme_images/timeseries_after.png)

#### Best of all, the functions are super flexible (and have a lot of cool features built into them!)
```python
from ahh import vis

vis_dict = dict(rows=2, # specify number of subplots
                sidebar_count=2, # specify number of side by side bars
                xlabel='x', title='Vis Demonstration', suptitle=True,
                title_pad=0.9, figsize='na') # set shared settings

vis.set_figsize(15, 8) # set figure size
vis.plot_bar(x, y, label='Red Bars', ylabel='Values', **vis_dict)
vis.plot_bar(x, y2, label='Blue Bars', sidebar_pos=2, color='blue', **vis_dict)

vis.plot_hist(['Chicken'] * 5 + ['Egg'] * 2 + ['Spam'], # random data
              rows=2, pos=2, ptype='bar',
              ylabel='Count', color='orange', # labels and color
              cumsum=True, save='vis_demonstration') # get cumulative count and save
```
![After Map](/readme_images/vis_demonstration.png)

#### But data exploration doesn't begin with visualizations! Check out ext.ahh() too!
```python
from ahh import ext

ext.ahh(ds=arr_ds, arr=x) # name=data
```

```
            Name: ds
          Length: 366
      Dimensions: (366, 73, 144)
   Unnested Type: <class 'numpy.float64'>
Overarching Type: <class 'numpy.ndarray'>
Minimum, Maximum: 188.290, 315.300
 Average, Median: 277.191, 282.670

Snippet of values:
[ 238.1  238.1  238.1  238.1  238.1 ...,  238.1  238.1  238.1  238.1  238.1]


            Name: arr
          Length: 15
      Dimensions: (15,)
   Unnested Type: <class 'numpy.int64'>
Overarching Type: <class 'numpy.ndarray'>
Minimum, Maximum: 1.000, 15.000
 Average, Median: 8.000, 8.000

Snippet of values:
[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15]
```

#### Interested in seeing more? There's way more in the [examples](https://github.com/ahuang11/ahh/tree/master/examples) page!

## AVAILABLE MODULES:
pre | sci | ext | era | vis | exp
--- | --- | --- | --- | --- | ---
pre-analysis | scientific | extraneous | era of time | visualization | experimenting

### Documentation here: https://ahuang11.github.io/ahh/
### Random, but awesome, tips: https://github.com/ahuang11/ahhsumtips

## KNOWN ISSUES AND HACKS:
- Installation of cartopy/basemap can be a bit tedious; easy way to install cartopy and basemap if you have Anaconda: 'conda install -c conda-forge cartopy' and 'conda install -c anaconda basemap'
- Updates to packages listed in requirements may break this package; recommend creating a Python 3 environment and install the specific package versions listed in requirements.txt
- In version 0.6.2, I accidentally bloated the repo so I decided to reset the repo and start semantic versioning beginning at v1.0.0; you may want to `rm -rf ahh` and `git clone https://github.com/ahuang11/ahh.git` again (or newly available, `pip install ahh`)!

changelog.md

## CHANGELOG:
### - v1.0.3
    - Add data files to wheel?
### - v1.0.2
    - Missing data folder in pip so this version fixes?
### - v1.0.1
    - Now available on pip!
### - v1.0.0
    - Accidentally bloated the repo in past version so deleted the old one and started anew
    - Begin semantic versioning
    - Cleaned up duplicate/unnecessary files
### -v0.6.2
    - Updated readme with something more enticing
    - Updated requirements for newer versions
    - Moved changelog from README.md to changelog.md
    - Added tests
    - Removed era.cdf2dt()
    - Added dt input to exp.arr_1d()
    - Added exception handling to functions in ext
    - Added lower input in pre.read_csv()
    - Added sci.get_norm() and renamed sci.regress to sci.get_regression()
    - Many under-the-hood cleanup and bug fixes in vis
    - Removed mpl.use('agg') in vis
    - Changed defaults in vis
    - Added suptitle in vis
    - Improved xarray compatibility in vis.plot_map()
    - Added vis.plot() to plot multiple lines in a shortcut method
    - Separated histogram functionality in plot_bar() to plot_hist()
    - Added vis.plot_cbar() to plot just a color bar
    - Renamed vis.set_color_bar() to vis.set_cbar()
    - Added vis.get_region_latlim()
    - Improved set_inherited algorithm
    - Set tight_layout as default in savefig()
    - Cleaned up sketches folder
### -v0.6.1
    - Update readme
    - New ext.get_ocean_mask() to mask out ocean/land
    - New sci.regress() to fit a regression line
    - NCL_CMAPS now supported for Python 2.7
    - vis.plot_map() now has central_longitude and mask_ocean/mask_land input
    - vis.plot_line() can now detect maximum ylim from previous lines
### -v0.6.0
    - Updated pre.wget_fi() and example
    - Added pre.gen_fi_list() and pre.wget_list() and examples
    - Fixed bug with era.clockit()
    - Allowed fourth argument with ext.parallelize()
    - Bug fixes with vis.plot_bar() histograms and example
### -v0.5.9
    - Changed input order of ext.round_to()
    - Changed default cmap for vis.plot_cmap()
    - Set tight_layout input to be 'auto' for vis.functions()
    - Use percentile to determine vmin and vmax in vis.plot_map()
    - Added couple exception handlings in vis.functions()
    - Better handling of color bar ticks, format, and contours in vis.plot_map()
### -v0.5.8
    - New era.td2dict() that breakdowns datetime.timedelta into more intuitive time units
    - Refactored era.clockit()
    - Added ticks input to vis.prettify_ax()
    - Revamped vis.set_date_ticks()
    - Modified vis.set_minor_ticks default color and fixed xlocator 'years' input
    - Modified defaults for vis.set_major_grid() linestyle
    - Modified vis.set_minor_grid() defaults
### -v0.5.7
    - Added ext.strip_2ws() which reduces two spaces to one space
    - Ported over NCL colormaps for use with vis.get_cmap() and updated example
    - New vis.prettify_bokeh() to be used on bokeh plots and example
    - Better handling of xarrays for vis.plot_map()
    - Added vis.set_inline_label() to add inline labels and built in vis.plot_line()
    - vis.set_figsize() now returns fig
    - Added vis.get_color_list() to get either a list of RGB tuples or hexcodes from cmap
    - Updated requirements.txt and setup.py
### -v0.5.6
    - Fixed elapsed seconds in era.clockit()
    - Bug fixes to contour/contour2/stipple in vis.plot_map() and examples made
    - Bug fix to vis.plot_heatmap() getting the max/min
    - vis.get_cmap() now more intelligent in choosing n
### -v0.5.5
    - Made 2010 as default year for era.jul2dt() and new method of reading the calendar
    - era.clockit() prints more straightforward time elapsed and new functionality to save, show, and return_dt
    - era.mth2terc() merged into era.dt2seas() with additional inputs to limit to four season or have rolling seasons
    - New era.shift_months() which allows shifting datetimes by input number of months
    - Bug fix in vis.set_date_ticks() to prevent freezes
    - era.dt2jul/jul2dt/dt2seas/dtnow_and_clockit/shift_months/time2dt() examples created
### -v0.5.4
    - ext.ahh() now handles xarrays proficiently and prints out average/median
    - Added ext.ahhh() which can pass as many unnamed variables as preferred, but no modifyable settings
    - Added ext.ahh_xr_check() used as a helper in ext.ahh/ahhh()
    - ext.glob() now prints out the original glob_str if it errors out
    - Format of color bars in vis.functions() are now intelligently chosen if not provided
    - Better handling of the tick_locs/interval of color bars
    - Intelligently choose the orientation of bar values
    - vis.set_figsize() now can easily accept width/height
    - vis.set_ax() is added to create plot axes easily
    - Fixed bug with vis.get_cmap()
### -v0.5.3
    - Hotfixed bugs from previous version (undefined kwargs) and added some error handling
    - Improved vis/set_axtext_AND_set_figtext examples
    - Added vis/get_cmap example
    - New functionality in vis.get_cmap()
    - Modified vis.set_latlons() and vis.set_major/minor_grid() defaults and scaling
### -v0.5.2
    - Made a part of pre.read_csv() into era.spawn_date_times()
    - Fixed exp.arr_1d() xy return
    - Revised ext.mkdir() to utilize built in os.makedirs()
    - Added km2mi input to sci.convert()
    - Added capability and example for mapping data to size or color in vis.plot_scatter()
    - Added length_scale inputs to enable/disable scaling of labels based on length in vis.functions()
    - Added **kwargs input to vis.functions() where applicable for more flexibility
    - Added more handling with vis.set_date_ticks()
    - Added fha/fva inputs to vis.set_figtext() and vis.set_axtext()
    - Added inheritance to vis.functions()
    - Added vis.get_cmap() used to discretize registered color maps and also create color maps from list of colors

ahh/era.py
from ahh.ext import MISC
from calendar import isleap
from netCDF4 import num2date
from dateutil.relativedelta import relativedelta
import matplotlib.dates as mdates
import xarray as xr
import pandas as pd
import numpy as np
import datetime
import os

__author__ = 'huang.andrew12@gmail.com'
__copyright__ = 'Andrew Huang'


class OutOfRange(Exception):
    pass


def dt2jul(dt):
    """
    Return julian day out of a datetime.

    :param dt: (datetime.datetime/int) - datetime
    :return jday: (int) - julian day
    """
    return dt.timetuple().tm_yday


def jul2dt(jday, year=2010):
    """
    Find the datetime from a julian date.

    :param jday: (int) - julian day
    :param year: (int) - year to determine if leap
    :return dt: (datetime.datetime) - respective datetime
    """
    this_dir = os.path.dirname(os.path.realpath(__file__))

    if isleap(year):
        if jday < 1 or jday > 366:
            print('\nInput Julian day is out of 1-366 range!\n')
            raise(OutOfRange)
        cal_str = 'jd_cal_leap'
    else:
        if jday < 1 or jday > 365:
            print('\nInput Julian day is out of 1-365 range!\n')
            raise(OutOfRange)
        cal_str = 'jd_cal'

    cal_path = os.path.join(this_dir, 'data', '{}.pkl'.format(cal_str))

    cal = pd.read_pickle(cal_path)
    jday_idc = np.where(cal == jday)

    mth = jday_idc[1][0] + 1
    day = jday_idc[0][0] + 1
    return datetime.datetime(year, mth, day)


def dtnow():
    """
    Get current UTC in datetime.

    :return utcnow: (datetime.datetime) - UTC now in datetime
    """
    return datetime.datetime.utcnow()


def td2dict(td):
    """
    Get a breakdown of the timedelta

    :param td: (datetime.timedelta) - datetime timedelta
    :return td_breakdown: (dict) - dictionary containing breakdown
    """
    d = td.days
    h = td.seconds // 3600
    m = (td.seconds // 60) % 60
    s = td.seconds % 60
    return dict(days=d, hours=h, minutes=m, seconds=s)


def clockit(start, n='', save=None, show=True, return_td=False):
    """
    Get elapsed time since start.

    :param start: (datetime.datetime) - start datetime
    :param n: (str) - label/description
    :param save: (str) - name of file to export error message to
    :param show: (boolean): whether to print time taken
    :param return_dt: (boolean): whether to return datetime timedelta taken
    :return time_taken: (datetime.timedelta) - time taken
    """
    td_taken = datetime.datetime.utcnow() - start
    td_dict = td2dict(td_taken)

    days_taken = td_dict['days']
    hours_taken = td_dict['hours']
    minutes_taken = td_dict['minutes']
    seconds_taken = td_dict['seconds']

    clockit_fmt = '{0} Days, {1} Hours, {2} Minutes, {3} Seconds Elapsed'
    clockit_fmtd = clockit_fmt.format(days_taken,
                                      hours_taken,
                                      minutes_taken,
                                      seconds_taken)

    if show:
        print(n)
        print(clockit_fmtd)
        print('')

    if save is not None:
        if '.txt' not in save:
            save += '.txt'
        with open(save, 'a') as f:
            f.write(clockit_fmtd)
            f.write('\n')

    if return_td:
        return td_taken


def time2dt(time=None, strf='infer', calendar='standard',
            year=None, month=None, day=None,
            hour=None, minute=None, second=None):
    """
    Creates a datetime array based on an unopened time variable or arrays of
    time strings.

    :param time: (netCDF4.Variable/str) - unopened time var/string type array
    :param strf: (str) - time format of string, option of 'infer'
    :param calendar: (str) - type of calendar
    :param year: (arr) - array of years
    :param month: (arr) - array of months
    :param day: (arr) - array of days
    :param hour: (arr) - array of hours
    :param minute: (arr) - array of minutes
    :param second: (arr) - array of seconds
    :return dt_arr: (np.array) - array of datetimes
    """
    if strf is not None:
        if strf is 'infer':
            dt_arr = pd.to_datetime(time, infer_datetime_format=True)
        else:
            dt_arr = pd.to_datetime(time, format=strf)
    elif year is not None:
        dt_dict = {'year': year, 'month': month, 'day': day}
        if hour is not None:
            dt_dict['hour'] = hour
        if minute is not None:
            dt_dict['minute'] = minute
        if second is not None:
            dt_dict['second'] = second
        dt_arr = pd.to_datetime(dt_dict)
    else:
        try:
            dt_arr = num2date(
                time[:], units=time.units, calendar=time.calendar)
        except:
            dt_arr = num2date(
                time[:], units=time.units, calendar=calendar)
    try:
        return pd.DatetimeIndex(dt_arr)
    except:
        print('Unable to return pd.DatetimeIndex, returning dt_arr')
        return dt_arr


def dt2seas(dts, four=True, target=None):
    """
    From a datetime array create an array of seasons.

    :param dts: (arr) - array of datetimes
    :param four: (boolean) - limit to four seasons
    :param target: (str) - whether the season "begin" or "end" at month
    :return dt_arr: (xr.DataArray) - array of seasons
    """
    dts = pd.DatetimeIndex(dts)

    if target is 'begin':
        dts = dts.shift(3, freq='1M')

    if four and not target:
        ds = xr.Dataset({'dts': dts})
        return ds.dts.dt.season.values
    else:
        month_in = MISC['months_initial']
        months = dts.month
        try:
            return np.array([(month_in[month - 3] +
                             month_in[month - 2] +
                             month_in[month - 1]) for
                             month in months])
        except:
            return (month_in[months - 3] +
                    month_in[months - 2] +
                    month_in[months - 1])


def dt2spec(dt, spec):
    """
    Convert datetimes to specified time value.

    :param dt: (datetime.datetime) - array of datetimes
    :param spec: (str) - name of time value
    :return spec_arr: (list) - list of specified time values
    """
    if isinstance(dt[0], datetime.timedelta):
        if spec is 'day':
            return [x.days for x in dt]
        elif spec is 'month':
            return [x.months for x in dt]
        elif spec is 'year':
            return [x.years for x in dt]
        elif spec is 'minute':
            return [x.minutes for x in dt]
        elif spec is 'second':
            return [x.seconds for x in dt]
    else:
        if spec is 'day':
            return [x.day for x in dt]
        elif spec is 'month':
            return [x.month for x in dt]
        elif spec is 'year':
            return [x.year for x in dt]
        elif spec is 'minute':
            return [x.minute for x in dt]
        elif spec is 'second':
            return [x.second for x in dt]


def str2spec(str_arr, spec):
    """
    Convert strings to specified time value assuming
    '%Y-%m-%d %H:%M:%S' format.

    :param str: (str) - array of strings
    :param spec: (str) - name of time value
    :return spec_arr: (np.array) - list of specified time values
    """
    if spec is 'day':
        return np.array([x[8:10] for x in str_arr], dtype='int')
    elif spec is 'month':
        return np.array([x[5:7] for x in str_arr], dtype='int')
    elif spec is 'year':
        return np.array([x[0:4] for x in str_arr], dtype='int')
    elif spec is 'hour':
        return np.array([x[11:13] for x in str_arr], dtype='int')
    elif spec is 'minute':
        return np.array([x[14:16] for x in str_arr], dtype='int')
    elif spec is 'second':
        return np.array([x[17:19] for x in str_arr], dtype='int')


def dt2num(dt, reverse=False):
    """
    Convert datetime to a number to indicate location on matplotlib plot.

    :param dt: (datetime.datetime) - a datetime object
    :param reverse: (boolean) - whether to reverse the operation
    :return mdate: (scalar) - value of datetime as a number
    """
    if reverse:
        return mdates.num2date(dt)
    else:
        return mdates.date2num(dt)


def spawn_dates_times(df, spawn_dates=True, spawn_times=False):
    """
    Build date/times column from a timeseries dataframe

    :param df: (pd.DataFrame) - dataframe with datetime index
    :param spawn_dates: (boolean) - whether to spawn year, month, day cols
    :param spawn_times: (boolean) - whether to spawn hour, minute, second cols
    :return df: (pd.DataFrame) - dataframe with datetime index
    """
    if spawn_dates:
        ind = df.index
        df = df.assign(year=ind.year, month=ind.month, day=ind.day)

    if spawn_times:
        ind = df.index
        df = df.assign(hour=ind.hour, minute=ind.minute, second=ind.second)

    return df


def shift_months(dts, nmonths):
    """
    Shift datetime(s) by number of months

    :param dts: (datetime.datetime/arr) - datetime objects
    :param nmonths: (int) - number of months to shift the datetime
    :return dts: (datetime.datetime) - datetime shifted by nmonths
    """
    try:
        return pd.DatetimeIndex([dt + relativedelta(months=nmonths)
                                 for dt in dts])
    except:
        return dts + relativedelta(months=nmonths)
